### ISSUE TITLE
feature: deterministic random_split

### ISSUE THREAD

**Issue by artemisart** [pytorch/pytorch#32467]
## 🚀 Feature
Option to provide a seed (random_state) for random_split() like the sklearn API https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html.

## Motivation

Useful for deterministic sampling & reproducible data generation (easily, without affecting the PRNG for other uses).

## Pitch / implementation proposal

In `dataset.py`
```diff
+ from torch.random import manual_seed, get_rng_state, set_rng_state

-def random_split(dataset, lengths):
+def random_split(dataset, lengths, random_state=None):
    r"""
    Randomly split a dataset into non-overlapping new datasets of given lengths.

    Arguments:
        dataset (Dataset): Dataset to be split
        lengths (sequence): lengths of splits to be produced
+       random_state (int, optional): seed used by the random number generator
    """
    if sum(lengths) != len(dataset):
        raise ValueError("Sum of input lengths does not equal the length of the input dataset!")

+   if random_state is not None:
+       state = get_rng_state()
+       manual_seed(random_state)

    indices = randperm(sum(lengths)).tolist()
-   return [Subset(dataset, indices[offset - length:offset]) for offset, length in zip(_accumulate(lengths), lengths)]
+   splits = [Subset(dataset, indices[offset - length:offset]) for offset, length in zip(_accumulate(lengths), lengths)]
+   if random_state is not None:
+       set_rng_state(state)
+   return splits
```

It could also allow a `ByteTensor` to be compatible with `get_rng_state()`.

## Alternatives

Manually call `manual_seed()` before splitting, without forgetting to save and restore the state with `get/set_rng_state()` or with `fork_rng()` (by the way it's not obvious from the docs that it should used with `with`, maybe add a snippet?)

**Comment by artemisart** [comment]
Should I make a PR for this feature or is it not wanted ?

**Comment by zplizzi** [comment]
I was just about to request the same thing. This would be really helpful - you want to be able to always get the same random split so that you can continue training a model after a crash, for example. Or so that you can later get access to the test set for evaluation of the trained model.

I would be fine with the alternative of setting+restoring RNG state if there was a code example provided, but as-is messing with RNG state is a great way to introduce subtle bugs so figuring it out myself is not something I'm too excited about.

**Comment by artemisart** [comment]
Merged in #34043 (@zplizzi)
